// react-progressive-bg-image/lib/Img.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _templateObject = _taggedTemplateLiteral(['\n  height: 100%;\n  background-repeat: no-repeat;\n  transition: ', ';\n\n  opacity: ', ';\n  filter: ', ';\n  /* this is needed so Safari keeps sharp edges */\n  transform: ', ';\n'], ['\n  height: 100%;\n  background-repeat: no-repeat;\n  transition: ', ';\n\n  opacity: ', ';\n  filter: ', ';\n  /* this is needed so Safari keeps sharp edges */\n  transform: ', ';\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _omit = require('ramda/src/omit');

var _omit2 = _interopRequireDefault(_omit);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var omitProps = (0, _omit2.default)(['blur', 'transition', 'isCached', 'isLoaded', 'opacity', 'scale', 'placeholder']);

var BaseComponent = function BaseComponent(_ref) {
  var component = _ref.component,
      children = _ref.children,
      otherProps = _objectWithoutProperties(_ref, ['component', 'children']);

  return _react2.default.createElement(component, omitProps(otherProps), children);
};
BaseComponent.displayName = 'BaseComponent';
BaseComponent.propTypes = {
  component: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.string]).isRequired,
  children: _propTypes2.default.node // Remind: There is not a children for Input tag.
};
var StyledImg = (0, _styledComponents2.default)(BaseComponent)(_templateObject, function (props) {
  return props.isCached ? 'none' : props.transition;
}, function (props) {
  return props.isLoaded ? 1 : props.opacity;
}, function (props) {
  return props.isLoaded ? 'none' : 'blur(' + props.blur + 'px)';
}, function (props) {
  return props.isLoaded ? 'none' : 'scale(' + props.scale + ')';
});

var Img = function Img(_ref2) {
  var component = _ref2.component,
      image = _ref2.image,
      style = _ref2.style,
      otherProps = _objectWithoutProperties(_ref2, ['component', 'image', 'style']);

  return _react2.default.createElement(StyledImg, Object.assign({}, otherProps, {
    component: component
  }, component === 'img' && { src: image }, {
    style: Object.assign({}, style, component !== 'img' && { backgroundImage: 'url("' + image + '")' })
  }));
};

Img.displayName = 'Img';
Img.propTypes = {
  // Internal
  image: _propTypes2.default.string.isRequired,
  isLoaded: _propTypes2.default.bool.isRequired,
  isCached: _propTypes2.default.bool.isRequired,

  // props
  opacity: _propTypes2.default.number.isRequired,
  blur: _propTypes2.default.number.isRequired,
  scale: _propTypes2.default.number.isRequired,
  className: _propTypes2.default.string,
  transition: _propTypes2.default.string,
  style: _propTypes2.default.object,
  component: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func])
};

exports.default = Img;
